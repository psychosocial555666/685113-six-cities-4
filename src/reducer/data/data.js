import {getOffersInCity, extend, getSortedOffers} from '../../utils';
import {SortType} from '../../const.js';

const initialState = {
  city: {
    cityName: `Amsterdam`,
    cityCoords: [52.38333, 4.9],
  },
  currentSortType: `Popular`,
  allOffers: [],
  currentOffers: [],
  currentComments: [],
};

const ActionType = {
  LOAD_OFFERS: `LOAD_OFFERS`,
  LOAD_COMMENTS: `LOAD_COMMENTS`,
  CHANGE_LOCATION: `CHANGE_LOCATION`,
  CHANGE_CURRENT_OFFERS: `CHANGE_CURRENT_OFFERS`,
  SORT_OFFERS: `SORT_OFFERS`,
};

const ActionCreator = {
  loadOffers: (offersAll) => {
    return {
      type: ActionType.LOAD_OFFERS,
      payload: offersAll,
    };
  },
  loadComments: (commentsAll) => {
    return {
      type: ActionType.LOAD_COMMENTS,
      payload: commentsAll,
    };
  },
  changeLocation: (city) => ({
    type: ActionType.CHANGE_LOCATION,
    payload: city,
  }),
  changeCurrentOffers: (cityName) => ({
    type: ActionType.CHANGE_CURRENT_OFFERS,
    payload: cityName,
  }),
  sortOffers: (sortType) => ({
    type: ActionType.SORT_OFFERS,
    payload: sortType,
  }),
};

const Operation = {
  loadOffers: (adaptCallback) => (dispatch, getState, api) => {
    return api.get(`/hotels`)
      .then((response) => {
        const data = adaptCallback ? adaptCallback(response.data) : response.data;
        dispatch(ActionCreator.loadOffers(data));
      });
  },
  loadComments: (adaptCallback, id) => (dispatch, getState, api) => {
    if (id) {
      return api.get(`/comments/${id}`)
      .then((response) => {
        const data = adaptCallback ? adaptCallback(response.data) : response.data;
        dispatch(ActionCreator.loadComments(data));
      });
    }
    return null;
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_OFFERS:
      return extend(state, {allOffers: action.payload,
        currentOffers: getOffersInCity(state.city.cityName, action.payload)}
      );
    case ActionType.LOAD_COMMENTS:
      return extend(state, {currentComments: action.payload});
    case ActionType.CHANGE_LOCATION:
      return extend(state, {city: action.payload});
    case ActionType.CHANGE_CURRENT_OFFERS:
      return extend(state, {
        currentOffers: getOffersInCity(action.payload, state.allOffers),
      });
    case ActionType.SORT_OFFERS:
      return extend(state, {
        currentSortType: action.payload,
        allOffers: action.payload === SortType.POPULAR
          ? state.allOffers
          : getSortedOffers(state.allOffers, action.payload),
      });
    default:
      return state;
  }
};

export {reducer, Operation, ActionType, ActionCreator};
