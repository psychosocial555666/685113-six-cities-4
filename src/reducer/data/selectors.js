import {createSelector} from "reselect";
import NameSpace from "../name-space.js";


export const getOffers = (state) => {
  return state[NameSpace.DATA].offers;
};

export const getCurrentCity = (state) => {
  return state[NameSpace.DATA].currentCity;
};

export const getCurrentCityName = (state) => {
  return state[NameSpace.DATA].currentCity.cityName;
};

export const getCurrentOffers = createSelector(
    getOffers,

    getCurrentCityName,

    (resultOne, resultTwo) => {
      return resultOne.filter((it) => it.cityName === resultTwo);
    }
);
