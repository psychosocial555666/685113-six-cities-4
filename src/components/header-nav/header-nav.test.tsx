import configureStore from "redux-mock-store";
import HeaderNav from "./header-nav";
import NameSpace from "../../reducer/name-space";
import * as React from "react";
import * as renderer from "react-test-renderer";
import thunk from 'redux-thunk';
import {Router} from "react-router-dom";
import {Operation} from '../../reducer/data/data';
import {Provider} from "react-redux";
import history from '../../history.js';


const middlewares = [thunk];
const mockStore = configureStore(middlewares);

jest.mock(`../../reducer/data/data`);
Operation.loadFavoriteOffers = () => (dispatch) => dispatch(jest.fn());
Operation.changeFavoriteStatus = () => (dispatch) => dispatch(jest.fn());

describe(`HeaderNav`, () => {
  it(`Should HeaderNav render correctly`, () => {
    const store = mockStore({
      [NameSpace.USER]: {
        userEmail: ``,
      }
    });

    const checkAuth = jest.fn();

    const tree = renderer
        .create(
            <Provider store={store}>
              <Router history={history} >
                <HeaderNav
                  checkAuth={checkAuth}
                />
              </Router>
            </Provider>, {createNodeMock: () => {
              return document.createElement(`div`);
            }})
        .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
