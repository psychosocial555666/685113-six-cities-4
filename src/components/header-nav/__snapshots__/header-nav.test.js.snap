// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderNav Should HeaderNav render correctly 1`] = `
<nav
  className="header__nav"
>
  <ul
    className="header__nav-list"
  >
    <li
      className="header__nav-item user"
    >
      <a
        className="header__nav-link header__nav-link--profile"
        href="/favorite"
        onClick={[Function]}
      >
        <div
          className="header__avatar-wrapper user__avatar-wrapper"
        />
        <span
          className="header__user-name user__name"
        >
          Sing in
        </span>
      </a>
    </li>
  </ul>
</nav>
`;
